src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -232,6 +476,31 @@ public int GetCount(bool onlyIfCheap)
 
                 return count;
             }
+
+            protected List<TSource> PopulateList(List<TSource> list)
+            {
+                for (int i = 0; ; i++)
+                {
+                    IEnumerable<TSource> source = GetEnumerable(i);
+                    if (source == null)
+                    {
+                        break;
+                    }
+
+                    list.AddRange(source);
+                }
+
+                return list;
+            }
+        }
+
+        private abstract class ConcatCollectionIterator<TSource> : ConcatIterator<TSource>
+        {
+            internal abstract int Count { get; }
```

Comment:
You should just override GetCount in the deriving implementations rather than introducing another virtual Count and overriding that.